### DISCUSSION TITLE
New event: mage_installed_exception (to i.e. forward exceptions to a 3rd party error logger, such as Sentry)

### DISCUSSION THREAD

**Discussion by pquerner** [OpenMage/magento-lts#3510] [Ideas]
### Description (*)

I'd like a new event (naming propose: mage_installed_exception) to catch events thrown before Mage sends the response to the frontend.

### Benefits

Using this new event ("mage_installed_exception") one can integrate their own way of handling Exceptions and i.e. forward them to a 3rd party error catcher, such as Sentry.

### Additional information

I could not find another way of catching or forwarding an (Unhandled?) exceptions, for instance when thrown from Observer context.
Exceptions are caught at "\Mage::run" and thats it.

I've tested integrating the Sentry 1.11 SDK (called Raven) in a OpenMage 20.0.12 enviroment.
I've installed Sentry SDK using composer into a custom modules directory and used the additional includes glob from the Mage.php file 

```
foreach (glob(BP . DS . 'app' . DS . 'etc' . DS . 'includes' . DS . '*.php') as $path) {
    include_once $path;
}
```

to auto-load a file called "sentry.init.php" with this content:

```
<?php

// Load composer autoload for lib sentry.
$vendorPathToSentry = MAGENTO_ROOT . DS . 'app' . DS . 'code' . DS . 'local' . DS . 'MyCustom'
    . DS . 'Module' . DS . 'lib' . DS . 'Sentry' . DS . 'vendor';
require_once $vendorPathToSentry . DS . 'autoload.php';

// TODO requireing composer autoload is not enough?!? mage autoloader puts himself in front and fucks me up or so oô
require_once $vendorPathToSentry . DS . 'sentry'
    . DS . 'sentry' . DS . 'lib' . DS . 'Raven' . DS . 'ErrorHandler.php';
require_once $vendorPathToSentry . DS . 'sentry'
    . DS . 'sentry' . DS . 'lib' . DS . 'Raven' . DS . 'Breadcrumbs' . DS . 'ErrorHandler.php';
Raven_Autoloader::register(); // tested with and without

```

Please not that I had to manually import files from Sentry nonetheless, because I think the autoloader from Magento doesnt find the files - although the computer autoloader should've caught that? Not sure.. But I reckon the Mage autoloader overwrites / destroys the other autoloaders.

Creating the new event allows to catch it in an observer, and forward the exception however you like.

For example here:

```
//... "run" method from Mage.php
    } catch (Exception $e) {
        if (self::isInstalled()) {
            self::dispatchEvent('mage_installed_exception', array('exception' => $e));    
            self::printException($e);
            exit();
    }
//...
```
```
/** @var Exception $exception */
$exception = $observer->getData('exception');

/** @var xxx $sentryClient */
$sentryClient = Mage::getModel('xxx/sentry_client'); // in the client you fetch your sentry settings from wherever you like (for instance settings.xml)
$sentryClient->sendException($exception); //forward to sentry
```

**Comment by kyrena** [comment]
Sorry for this ads like, for Sentry, you can use [openmage-sentry](https://github.com/luigifab/openmage-sentry).

**Reply by pquerner** [reply]
That module changes too many unessesary files - in contrast to only 1 simply event.
And uses an outdated Sentry Library (internal 0.1 something, while 1.11 exists for <= PHP 7.1)

So not sure about that one.
I think I like my current solution more which just depends on adding a simply "dispatchEvent" more.

Although I must say I havent tested all the cases yet (sql errors, frontend errors, ...)

**Comment by sreichel** [comment]
Maybe worth to test ... https://github.com/firegento/firegento-logger/tree/develop/src/lib/sentry

**Reply by pquerner** [reply]
That only catches `Mage::log`.

**Reply by m-overlund** [reply]
@luigifab also have a Sentry implementation
https://github.com/luigifab/openmage-sentry

**Reply by pquerner** [reply]
That was already proposed here.
https://github.com/OpenMage/magento-lts/discussions/3510#discussioncomment-6989714

I didnt want this discussion to become a "how to implement sentry". I am just saying, that its currently not possible to handle uncatched exceptions (to my knowledge!), which will only be handled once in `Mage::run` and a new event here helps with that.

The sentry implementation is probably fine in either modules.

**Comment by pquerner** [comment]
Just want to update you that https://github.com/OpenMage/magento-lts/pull/3613 exists.
Its included in our fork of OpenMage for a while now and didnt result in any failures.

If anyone is interested in my implementation of Sentry, let me know and I'll write it down in some bullet points. 
(Its a little bit different than the mentioned solutions in this thread)
